cartography/intel/gsuite/api.py

```diff
@@ -0,0 +1,243 @@
+        try:
+            resp = request.execute()
+        except HttpError as e:
+            logger.warning('HttpError occurred in api.get_groups_for_email(), returning empty list. Details: %r', e)
+            groups = []
+            break
+        groups = groups + resp.get('groups', [])
+        request = admin.groups().list_next(request, resp)
+    return groups
+
+
+def get_members_for_group(admin, group_email):
+    """ Get all members for a google group
+
+    :param group_email: A string representing the email address for the group
+
+    :return: List of dictionaries representing Users or Groups.
+    """
+    request = admin.members().list(
+        groupKey=group_email,
+        maxResults=500,
+    )
+    members = []
+    while request is not None:
+        try:
+            resp = request.execute()
+        except HttpError as e:
+            logger.warning('HttpError occurred in api.get_members_for_group(), returning empty list. Details: %r', e)
+            members = []
+            break
+        members = members + resp.get('members', [])
+        request = admin.members().list_next(request, resp)
+
+    return members
+
+
+def get_all_users(admin):
+    """
+    Return list of Google Users in your organization
+    Returns empty list if we are unable to enumerate the groups for any reasons
+    https://developers.google.com/admin-sdk/directory/v1/guides/manage-users
+
+    :param admin: apiclient discovery resource object
+    see
+    :return: List of Google users in domain
+    see https://developers.google.com/admin-sdk/directory/v1/guides/manage-users#get_all_domain_users
+    """
+    request = admin.users().list(customer='my_customer', maxResults=500, orderBy='email')
+    users = []
+    while request is not None:
+        try:
+            resp = request.execute()
+        except HttpError as e:
+            logger.warning('HttpError occurred in api.get_all_users(), returning empty list. Details: %r', e)
+            users = []
+            break
+        users = users + resp.get('users', [])
+        # break
+        request = admin.users().list_next(request, resp)
+    return users
+
+
+def get_ingestion_groups_qry():
+    return """
+        UNWIND {GroupData} as group
+        MERGE (g:GSuiteGroup{id: group.id})
+        ON CREATE SET
+        g.admin_created = group.adminCreated,
+        g.description = group.description,
+        g.direct_members_count = group.directMembersCount,
+        g.email = group.email,
+        g.etag = group.etag,
+        g.kind = group.kind,
+        g.name = group.name,
+        g.lastupdated = {UpdateTag}
+    """
+
+
+def load_gsuite_groups(session, groups, gsuite_update_tag):
+    logger.info('Ingesting {} gsuite groups'.format(len(groups)))
+    session.run(get_ingestion_groups_qry(), GroupData=groups, UpdateTag=gsuite_update_tag)
+
+
+def get_ingestion_users_qry():
+    return """
+        UNWIND {UserData} as user
+        MERGE (u:GSuiteUser{id: user.id})
```

Comment:
Is this ID guaranteed to be unique?  What does it look like?

Reply:
the ID generated by Google is a unique id.  It's length 15, lower case alphanumeric.  ex: `o36w3mzh00fkzbq`

Reply:
Cool, that works.